Sources/DBusSwift/DBusAuthenticationHandler.swift

```diff
@@ -0,0 +1,171 @@
+/// See: https://dbus.freedesktop.org/doc/dbus-specification.html#auth-protocol
+public final class DBusAuthenticationHandler: ChannelDuplexHandler, @unchecked Sendable {
+  public typealias InboundIn = ByteBuffer
+  public typealias InboundOut = ByteBuffer
+  public typealias OutboundIn = ByteBuffer
+  public typealias OutboundOut = ByteBuffer
+
+  /// States of the DBus authentication protocol
+  /// See: https://dbus.freedesktop.org/doc/dbus-specification.html#auth-protocol
+  public enum State {
+    /// Waiting for NUL byte reply from server (initial state)
+    case waitingForNullReply
+    /// Authentication sent, waiting for OK response
+    case waitingForOK
+    /// Successfully authenticated, normal message passing can begin
+    case authenticated
+    /// Authentication failed
+    case failed
+  }
+
+  private var state: State = .waitingForNullReply
+  private var buffer = ByteBufferAllocator().buffer(capacity: 128)
+  private let auth: AuthType
+  private var writeBuffer = [ByteBuffer]()
+
+  public init(auth: AuthType) {
+    self.auth = auth
+  }
+
+  public func channelRead(context: ChannelHandlerContext, data: NIOAny) {
+    var buf = self.unwrapInboundIn(data)
+    buffer.writeBuffer(&buf)
+    processBuffer(context: context)
+  }
+
+  private func processBuffer(context: ChannelHandlerContext) {
+    defer { buffer.discardReadBytes() }
+    while buffer.readableBytes > 0 {
+      switch state {
+      case .waitingForNullReply:
+        // Wait for server's NUL byte reply (optional, but some servers send it)
+        if let byte = buffer.getInteger(at: buffer.readerIndex, as: UInt8.self), byte == 0 {
+          buffer.moveReaderIndex(forwardBy: 1)
+        }
+
+        state = .waitingForOK
+      case .waitingForOK:
+        guard var line = buffer.readString(length: buffer.readableBytes) else { return }
+        if line.starts(with: "OK ") {
+          line = String(line.dropFirst(3))
+
+          let begin = "BEGIN\r\n"
+          let out = context.channel.allocator.buffer(string: begin)
+          context.writeAndFlush(self.wrapOutboundOut(out), promise: nil)
+
+          do {
+            let handler = try context.pipeline.syncOperations.handler(
+              type: ByteToMessageHandler<LineBasedFrameDecoder>.self)
+            _ = context.pipeline.syncOperations.removeHandler(handler)
+
+            // Directly process buffers after handler removal
+            for buffer in self.writeBuffer {
+              context.writeAndFlush(self.wrapOutboundOut(buffer), promise: nil)
+            }
+            self.writeBuffer.removeAll(keepingCapacity: true)
+            self.state = .authenticated
+            context.fireChannelActive()
+            context.fireChannelWritabilityChanged()
+          } catch {
+            context.fireErrorCaught(error)
+          }
+        } else if line.starts(with: "REJECTED ") {
+          // let mechanisms = line.split(separator: " ")
+          //     .dropFirst()
+
+          return
```

Comment:
`context.fireErrorCaught` should be done, or we should retry authentication with a different mechanism